### PR TITLE
Use ArgparsedCommand for everything

### PR DESCRIPTION
**Pull request by stnevans**
As per #244, we want to migrate to use ArgparsedCommand.
This pr currently contains changes to move (almost) all commands to ArgparsedCommands.

It's not quite ready to be merged yet. I want to make sure everything is tested well before merging, as any typos could easily break a command. 

In addition, there are a few places where I think the description I gave could be improved. Feel free to suggest changes for any of them.

I also left a couple of TODOs in the code that I would appreciate input on.

Finally, I wasn't sure what to do with QuietSloppyParsedCommands.

### CODE REVIEW

File: pwndbg/commands/argv.py [hunk 1]
```diff
@@ -13,20 +15,18 @@
...
-    Prints out the number of arguments.
-    """
     print(pwndbg.argv.argc)
 
-@pwndbg.commands.ParsedCommand
+
+parser = argparse.ArgumentParser()
+parser.description = "Prints out the contents of argv."
+parser.add_argument("i", nargs='?', type=int, default=None, help="Index of the argument to print out.")
+@pwndbg.commands.ArgparsedCommand(parser,aliases=["args"])
```
Comment (disconnect3d) [inline pwndbg/commands/argv.py:27]:
```suggestion
@pwndbg.commands.ArgparsedCommand(parser,  aliases=["args"])
```

File: pwndbg/commands/argv.py [hunk 2]
```diff
@@ -36,17 +36,16 @@ def argv(i=None):
...
-@pwndbg.commands.Command
+parser = argparse.ArgumentParser()
+parser.description = "Prints out the contents of the environment."
+parser.add_argument("name", nargs='?', type=str, default=None, help="Name of the environment variable to see.")
+@pwndbg.commands.ArgparsedCommand(parser,aliases=["env","environ"])
 @pwndbg.commands.OnlyWhenRunning
 def envp(name=None):
+    if name is not None:
+        gdb.execute('p $environ("%s")' % name)
+        return
```
Comment (disconnect3d) [inline pwndbg/commands/argv.py:48]: We should delete the leftover function docstring that is below.

Comment (disconnect3d) [reply pwndbg/commands/argv.py:48]: Offtopic comment: the python api's `environ` function has meh error handling:
```
pwndbg> p $environ("PWD=")
Python Exception <class 'TypeError'> Could not convert Python object: void.:
Error while executing Python code.
```
We might want to send a patch to GDB to fix that :P

File: pwndbg/commands/argv.py [hunk 3]
```diff
@@ -36,17 +36,16 @@ def argv(i=None):
...
-    """
-    Prints out the contents of argv.
-    """
-    argv()
 
-@pwndbg.commands.Command
+parser = argparse.ArgumentParser()
+parser.description = "Prints out the contents of the environment."
+parser.add_argument("name", nargs='?', type=str, default=None, help="Name of the environment variable to see.")
+@pwndbg.commands.ArgparsedCommand(parser,aliases=["env","environ"])
```
Comment (disconnect3d) [inline pwndbg/commands/argv.py:43]:
```suggestion
@pwndbg.commands.ArgparsedCommand(parser,aliases=["env", "environ"])
```

File: pwndbg/commands/argv.py [hunk 4]
```diff
@@ -36,17 +36,16 @@ def argv(i=None):
...
-def args():
-    """
-    Prints out the contents of argv.
-    """
-    argv()
 
-@pwndbg.commands.Command
+parser = argparse.ArgumentParser()
+parser.description = "Prints out the contents of the environment."
+parser.add_argument("name", nargs='?', type=str, default=None, help="Name of the environment variable to see.")
```
Comment (disconnect3d) [inline pwndbg/commands/argv.py:42]:
```suggestion
parser.add_argument("name", nargs='?', type=str, default=None, help="Environment variable to display.")
```

Comment (disconnect3d) [reply pwndbg/commands/argv.py:42]: I'm okay with both versions

File: pwndbg/commands/context.py [hunk 5]
```diff
@@ -82,14 +83,22 @@ def output():
         return open(str( config_output ), "w")
 
 # @pwndbg.events.stop
```
Comment (disconnect3d) [inline pwndbg/commands/context.py:85]: Can we remove this comment?

File: pwndbg/commands/context.py [hunk 6]
```diff
@@ -82,14 +83,22 @@ def output():
...
-def context(*args):
+def context(subcontext=None):
     """
     Print out the current register, instruction, and stack context.
 
     Accepts subcommands 'reg', 'disasm', 'code', 'stack', 'backtrace', and 'args'.
     """
+    if subcontext is None:
+        subcontext = []
+    args = subcontext
```
Comment (disconnect3d) [inline pwndbg/commands/context.py:100]: Or just:
```python
    args = subcontext or config_context_sections.split()
```
and we could also remove the latter  `if len(args) == 0` and some code above

File: pwndbg/commands/context.py [hunk 7]
```diff
@@ -115,10 +124,12 @@ def context(*args):
 def context_regs():
     return [pwndbg.ui.banner("registers")] + get_regs()
 
-
-@pwndbg.commands.Command
+parser = argparse.ArgumentParser()
+parser.description = '''Print out all registers and enhance the information.'''
```
Comment (disconnect3d) [inline pwndbg/commands/context.py:128]:
```suggestion
parser.description = 'Print out all registers and enhance the information.'
```

File: pwndbg/commands/dt.py [hunk 8]
```diff
@@ -5,15 +5,24 @@
...
+    Dump out information on a type (e.g. ucontext_t).
 
-@pwndbg.commands.Command
+    Optionally overlay that information at an address.
+    """
+parser.add_argument("typename", type=str, help="The name of the structure being dumped.")
+parser.add_argument("address", type=int, nargs="?", default=None, help="The address of the structure.")
+@pwndbg.commands.ArgparsedCommand(parser)
 def dt(typename, address=None):
     """
```
Comment (disconnect3d) [inline pwndbg/commands/dt.py:27]: We should delete the docstring

File: pwndbg/commands/heap.py [hunk 9]
```diff
@@ -79,7 +79,11 @@ def format_bin(bins, verbose=False, offset=None):
 
     return result
 
-@pwndbg.commands.ParsedCommand
+
+parser = argparse.ArgumentParser()
+parser.description = "Prints out chunks starting from the address specified by `addr`."
+parser.add_argument("addr", nargs="?", type=int, default=None, help="The address of the heap.")
+@pwndbg.commands.ArgparsedCommand(parser)
```
Comment (disconnect3d) [inline pwndbg/commands/heap.py:86]: We should delete the function docstring

**Review by disconnect3d — COMMENTED** [review]
Reviewed a part of the code, see comments :)

File: pwndbg/commands/context.py (hunk 5 above)
Comment (stnevans) [reply pwndbg/commands/context.py:85]: Looks like this was added a long time ago. Should be fine to remove.

File: pwndbg/commands/context.py (hunk 7 above)
Comment (stnevans) [reply pwndbg/commands/context.py:128]: Looks like a good change, but for future reference, how much does this matter?
Is it just better style to use one quote if the string is just one line?

File: pwndbg/commands/heap.py (hunk 9 above)
Comment (stnevans) [reply pwndbg/commands/heap.py:86]: I wasn't sure if it was worth leaving them. If you think it makes sense, I will add a commit deleting all the doc strings.

File: pwndbg/commands/context.py (hunk 7 above)
Comment (disconnect3d) [reply pwndbg/commands/context.py:128]: > how much does this matter?
It doesn't; tho it is good to keep a proper coding style.

If you haven't seen pep8 it's probably good to read that https://www.python.org/dev/peps/pep-0008/ (and also take some of the rules with a grain of salt).

**Review by disconnect3d — APPROVED** [review]

**Comment by disconnect3d** [comment]
We can resolve other minor things like remove unnecesary docstrings and fix the style later on

**Pull request merged**